### PR TITLE
MYST-148 gracefully close on Interrupt, sigterm and sighup

### PR DESCRIPTION
**Pull request by shroomist**
is this enough for graceful termination, or do we need a more elaborate system?

### CODE REVIEW

File: cmd/mysterium_client/run/command.go [hunk 1]
```diff
@@ -52,11 +55,25 @@ func NewCommandWith(
...
 	httpApiServer := tequilapi.NewServer(options.TequilapiAddress, options.TequilapiPort, router)
-
-	return &CommandRun{
+	cmd := &CommandRun{
 		connectionManager,
 		httpApiServer,
 	}
+	sigterm := make(chan os.Signal, 2)
+	signal.Notify(sigterm, os.Interrupt, syscall.SIGTERM, syscall.SIGHUP)
+	go sarahConnor(sigterm, cmd)
```
Comment (Waldz) [inline cmd/mysterium_client/run/command.go]: Factory should be responsible 1 responsibility - for factoring, but not running

Comment (shroomist) [reply cmd/mysterium_client/run/command.go]: alright, this is now in Run()

File: cmd/mysterium_client/run/command.go [hunk 2]
```diff
@@ -67,6 +70,10 @@ type CommandRun struct {
 
 //Run starts Tequilapi service - does not block
 func (cmd *CommandRun) Run() error {
+	sigterm := make(chan os.Signal, 2)
```
Comment (donce) [inline cmd/mysterium_client/run/command.go]: Why `2`, isn't `1` enough?

File: cmd/mysterium_client/run/command.go [hunk 3]
```diff
@@ -98,3 +105,13 @@ func (cmd *CommandRun) Kill() error {
...
+
+func sigtermListener(terminator chan os.Signal, cmd *CommandRun) {
+	<-terminator
+	err := cmd.Kill()
+	if err != nil {
+		fmt.Printf("Unable to kill one of subroutines %q\n", err.Error())
+	}
+	fmt.Println("Good bye")
+	os.Exit(1)
+}
```
Comment (donce) [inline cmd/mysterium_client/run/command.go]: Closing subroutines and exiting is already implemented in `mysterium.client.go/newStopHandler` - can we re-use that somehow?

File: cmd/mysterium_client/run/command.go [hunk 4]
```diff
@@ -98,3 +105,13 @@ func (cmd *CommandRun) Kill() error {
 
 	return nil
 }
+
+func sigtermListener(terminator chan os.Signal, cmd *CommandRun) {
```
Comment (donce) [inline cmd/mysterium_client/run/command.go]: Function should be named as a verb, not as a noun - maybe `listenSigterm` suites better?

File: cmd/mysterium_client/run/command.go [hunk 5]
```diff
@@ -98,3 +105,13 @@ func (cmd *CommandRun) Kill() error {
...
+	<-terminator
+	err := cmd.Kill()
+	if err != nil {
+		fmt.Printf("Unable to kill one of subroutines %q\n", err.Error())
```
Comment (donce) [inline cmd/mysterium_client/run/command.go]: `os.Stderr` should be re-used for consistency, just like this:
```go
fmt.Fprintln(os.Stderr, err)
```

File: cmd/mysterium_client/run/command.go [hunk 6]
```diff
@@ -98,3 +105,13 @@ func (cmd *CommandRun) Kill() error {
...
+	}
+	fmt.Println("Good bye")
+	os.Exit(1)
```
Comment (Waldz) [inline cmd/mysterium_client/run/command.go]: `Exit(1)` mean program ending with errorNo=1, success graceful stop  should end with errorNo=0

File: cmd/mysterium_client/run/command.go (hunk 3 above)
Comment (donce) [reply cmd/mysterium_client/run/command.go]: It's hooked up differently, but the action it does on exit is duplicated:
```go
err := cmd.Kill()
if err != nil {
	fmt.Printf("Unable to kill one of subroutines %q\n", err.Error())
}
fmt.Println("Good bye")
os.Exit(1)
```

```go
if err := cmdRun.Kill(); err != nil {
	return err
}

os.Exit(0)
return nil
```

What bothers me is that these two duplications already differs - one returns exit `1` and prints outs "Good bye", another returns `0` without text.

We can easily DRY these - to move them to a separate method on `CommandRun` or to extend `CommandRun.Kill` with these duplications.

File: cmd/mysterium_client/run/command.go (hunk 6 above)
Comment (donce) [reply cmd/mysterium_client/run/command.go]: agree - we are already doing `os.Exit(0)` in `mysterium_client.go`

File: cmd/mysterium_server/command_run/command.go [hunk 7]
```diff
@@ -110,8 +110,15 @@ func (cmd *CommandRun) Wait() error {
 	return cmd.vpnServer.Wait()
 }
 
-func (cmd *CommandRun) Kill() {
+func (cmd *CommandRun) Kill() (err error) {
```
Comment (donce) [inline cmd/mysterium_server/command_run/command.go]: You're not really using named result variable - you can just leave `error`.

File: cmd/terminator.go [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+func NewTerminator(command Command) {
+	sigterm := make(chan os.Signal)
+	signal.Notify(sigterm, os.Interrupt, syscall.SIGTERM, syscall.SIGHUP)
+
+	go waitTerminatorSignals(sigterm, command)
+}
+
+func waitTerminatorSignals(terminator chan os.Signal, command Command) {
```
Comment (shroomist) [inline cmd/terminator.go]: for those who don't know -- this function is Sarah Connor

**Comment by shroomist** [comment]
@donce, is it better now?

**Pull request merged**